flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecWindowAggregate.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.flink.table.runtime.operators.window.slicing.SliceAssigner;
+import org.apache.flink.table.runtime.operators.window.slicing.SliceAssigners;
+import org.apache.flink.table.runtime.operators.window.slicing.SliceSharedAssigner;
+import org.apache.flink.table.runtime.types.LogicalTypeDataTypeConverter;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.tools.RelBuilder;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Stream {@link ExecNode} for window table-valued based aggregate.
+ *
+ * <p>The differences between {@link StreamExecWindowAggregate} and {@link
+ * StreamExecGroupWindowAggregate} is that, this node is translated from window TVF syntax, but the
+ * other is from the legacy GROUP WINDOW FUNCTION syntax. In the long future, {@link
+ * StreamExecGroupWindowAggregate} will be dropped.
+ */
+public class StreamExecWindowAggregate extends ExecNodeBase<RowData>
+        implements StreamExecNode<RowData> {
+
+    private static final long WINDOW_AGG_MEMORY_RATIO = 100;
+
+    private final int[] grouping;
+    private final AggregateCall[] aggCalls;
+    private final WindowingStrategy windowing;
+    private final PlannerNamedWindowProperty[] namedWindowProperties;
+
+    public StreamExecWindowAggregate(
+            int[] grouping,
+            AggregateCall[] aggCalls,
+            WindowingStrategy windowing,
+            PlannerNamedWindowProperty[] namedWindowProperties,
+            InputProperty inputProperty,
+            RowType outputType,
+            String description) {
+        super(Collections.singletonList(inputProperty), outputType, description);
+        this.grouping = grouping;
+        this.aggCalls = aggCalls;
+        this.windowing = windowing;
+        this.namedWindowProperties = namedWindowProperties;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Transformation<RowData> translateToPlanInternal(PlannerBase planner) {
+        final ExecEdge inputEdge = getInputEdges().get(0);
+        final Transformation<RowData> inputTransform =
+                (Transformation<RowData>) inputEdge.translateToPlan(planner);
+        final RowType inputRowType = (RowType) inputEdge.getOutputType();
+
+        final TableConfig config = planner.getTableConfig();
+        final SliceAssigner sliceAssigner = createSliceAssigner(windowing);
+
+        // Hopping window requires additional COUNT(*) to determine whether to register next timer
+        // through whether the current fired window is empty, see SliceSharedWindowAggProcessor.
+        final boolean needInputCount = sliceAssigner instanceof SliceAssigners.HoppingSliceAssigner;
```

Comment:
Could you please explain why need COUNT(*) for HoppingSliceAssigner in detail, I'm still confused about this. Thanks a lot.

Reply:
The COUNT(*) is used to know how much records in a window, similar to `RecordCounter` in `GroupAggFunction`. If the COUNT(*) of a window is `0`, then we don't need to register next timer for the hop window, otherwise, the timers for the hop window will grow unlimitly.

Reply:
Thanks for explain.